aws/resource_aws_dx_gateway_association.go

```diff
@@ -27,6 +27,9 @@ func resourceAwsDxGatewayAssociation() *schema.Resource {
 			State: resourceAwsDxGatewayAssociationImport,
 		},
 
+		SchemaVersion: 1,
+		MigrateState:  resourceAwsDxGatewayAssociationMigrateState,
```

Comment:
State migrations in Terraform 0.12 require using `StateUpgraders` instead of `MigrateState`. Please see the [Extending Terraform documentation](https://www.terraform.io/docs/extend/resources/state-migration.html#terraform-v0-12-sdk-state-migrations) for more information.

I verified this requirement with the following:

```hcl
# Terraform 0.11 syntax for easy flipping between 0.11.14 and 0.12.12
terraform {
  required_version = "0.12.12"
}

provider "aws" {
  region  = "us-east-2"
  version = "2.7.0"
}

resource "aws_dx_gateway" "test" {
  name            = "bflad-testing"
  amazon_side_asn = 65534
}

resource "aws_vpc" "test" {
  cidr_block = "10.255.255.0/28"

  tags = {
    Name = "bflad-testing"
  }
}

resource "aws_vpn_gateway" "test" {
  tags = {
    Name = "bflad-testing"
  }
}

resource "aws_vpn_gateway_attachment" "test" {
  vpc_id         = "${aws_vpc.test.id}"
  vpn_gateway_id = "${aws_vpn_gateway.test.id}"
}

resource "aws_dx_gateway_association" "test" {
  dx_gateway_id  = "${aws_dx_gateway.test.id}"
  vpn_gateway_id = "${aws_vpn_gateway_attachment.test.vpn_gateway_id}"
}
```

```console
$ terraform apply
# Update provider version constraint to 2.33.0
$ terraform init
$ terraform plan
# Note DirectConnectClientException
# Copy in provider from this branch (e.g. make build) and reinitialize
$ cp ~/go/bin/terraform-provider-aws .terraform/plugins/darwin_amd64/terraform-provider-aws_v2.33.0_x4
$ terraform init
$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

aws_dx_gateway.test: Refreshing state... [id=1bf32e9e-0dfc-4782-b7cf-5d94c5f719ef]
aws_vpn_gateway.test: Refreshing state... [id=vgw-0c0dad070ebae48ff]
aws_vpc.test: Refreshing state... [id=vpc-0c15c4266f46c1c99]
aws_vpn_gateway_attachment.test: Refreshing state... [id=vpn-attachment-8b7e34d3]
aws_dx_gateway_association.test: Refreshing state... [id=ga-1bf32e9e-0dfc-4782-b7cf-5d94c5f719efvgw-0c0dad070ebae48ff]

Warning: "vpn_gateway_id": [DEPRECATED] use 'associated_gateway_id' argument instead

  on main.tf line 34, in resource "aws_dx_gateway_association" "test":
  34: resource "aws_dx_gateway_association" "test" {



Error: error reading Direct Connect gateway association (ga-1bf32e9e-0dfc-4782-b7cf-5d94c5f719efvgw-0c0dad070ebae48ff): DirectConnectClientException: Association ID  has an invalid format.
	status code: 400, request id: 7d6ed9db-6e00-44a7-9ade-28d84341a5cf
```